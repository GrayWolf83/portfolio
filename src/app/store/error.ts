import { AppDispatch, RootState } from './index'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'

type errorState = {
	error: string | null
}

const initialState: errorState = {
	error: null,
}

const errorSlice = createSlice({
	name: 'errors',
	initialState,
	reducers: {
		loadingError(state, action: PayloadAction<string>) {
			state.error = action.payload
		},
		errorCleared(state) {
			state.error = null
		},
	},
})

const { loadingError, errorCleared } = errorSlice.actions

export const setLoadingError =
	(error: any) => async (dispatch: AppDispatch) => {
		if (error?.response?.data?.message) {
			dispatch(loadingError(error.response.data.message))
		}
	}

export const clearErrors = () => (dispatch: AppDispatch) => {
	dispatch(errorCleared())
}

export const getError = () => (state: RootState) => {
	return state.errors.error
}

export default errorSlice.reducer
