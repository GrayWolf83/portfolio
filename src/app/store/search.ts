import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type SearchState = {
	entities: IUser[]
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	page: number
	search: string
	usersCount: number
}

const initialState: SearchState = {
	entities: [],
	isLoading: false,
	error: null,
	dataLoaded: false,
	page: 1,
	search: '',
	usersCount: 0,
}

const searchSlice = createSlice({
	name: 'search',
	initialState,
	reducers: {
		serachLoadingStart: (state) => {
			state.isLoading = true
		},
		serachLoadingEnd: (state) => {
			state.isLoading = false
		},
		searchDataLoaded: (
			state,
			action: PayloadAction<{
				entities: IUser[]
				usersCount: number
				search: string
				page: number
			}>,
		) => {
			state.entities = action.payload.entities
			state.dataLoaded = true
			state.page = action.payload.page
			state.usersCount = action.payload.usersCount
			state.search = action.payload.search
		},
	},
})

const { serachLoadingStart, serachLoadingEnd, searchDataLoaded } =
	searchSlice.actions

export const searchUsersList =
	(search: string, page: number) => async (dispatch: AppDispatch) => {
		dispatch(serachLoadingStart())
		try {
			const payload = await userService.searchUsers(search, page)
			dispatch(
				searchDataLoaded({
					entities: payload.items,
					usersCount: payload.total_count,
					search,
					page,
				}),
			)
		} catch (error: any) {
			dispatch(setLoadingError(error))
		} finally {
			dispatch(serachLoadingEnd())
		}
	}

export const getUsersList = () => (state: RootState) => {
	return state.search.entities
}

export const getUsersCount = () => (state: RootState) => {
	return state.search.usersCount
}

export const getSearchLoadingStatus = () => (state: RootState) => {
	return state.search.isLoading
}

export const getSearchCurrentPage = () => (state: RootState) => {
	return state.search.page
}

export default searchSlice.reducer
