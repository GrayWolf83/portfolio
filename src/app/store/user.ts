import { IRepo } from '../types/IRepo'
import { AppDispatch, RootState } from './index'
import { setLoadingError } from './error'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../types/IUser'
import userService from '../services/user.service'

type userState = {
	currentUser: IUser | null
	isLoading: boolean
	error: string | null
	dataLoaded: boolean
	repos: IRepo[]
	skills: string[]
}

const initialState: userState = {
	currentUser: null,
	isLoading: false,
	error: null,
	dataLoaded: false,
	repos: [],
	skills: [],
}

const userSlice = createSlice({
	name: 'user',
	initialState,
	reducers: {
		userLoadingStart: (state) => {
			state.isLoading = true
		},
		userLoadingEnd: (state) => {
			state.isLoading = false
		},
		userLoadingError: (state, action: PayloadAction<string>) => {
			state.error = action.payload
		},
		userLoaded: (state, action: PayloadAction<IUser>) => {
			state.currentUser = action.payload
			state.dataLoaded = true
		},
		userReposLoaded: (state, action: PayloadAction<IRepo[]>) => {
			state.repos = action.payload
		},
		userSkillsLoaded: (
			state,
			action: PayloadAction<{ [key: string]: number }[]>,
		) => {
			let skills = action.payload
				.map((item) => {
					return Object.keys(item).map((name) => name)
				})
				.flat()

			state.skills = Array.from(new Set(skills))
		},
	},
})

const {
	userLoadingStart,
	userLoadingEnd,
	userLoaded,
	userReposLoaded,
	userSkillsLoaded,
	userLoadingError,
} = userSlice.actions

export const loadUser = (login: string) => async (dispatch: AppDispatch) => {
	dispatch(userLoadingStart())
	try {
		const payload = await userService.getUser(login)
		dispatch(userLoaded(payload))

		const repos = await userService.getRepos(login)
		dispatch(userReposLoaded(repos))
	} catch (error: any) {
		dispatch(userLoadingError(error))
	} finally {
		dispatch(userLoadingEnd())
	}
}

export const loadUserSkills =
	(names: string[]) => async (dispatch: AppDispatch) => {
		dispatch(userLoadingStart())
		try {
			const skills = await userService.getSkills(names)
			dispatch(userSkillsLoaded(skills))
		} catch (error: any) {
			dispatch(userLoadingError(error))
		} finally {
			dispatch(userLoadingEnd())
		}
	}

export const getCurrentUser = () => (state: RootState) => {
	return state.user.currentUser
}

export const getUserRepos = () => (state: RootState) => {
	return state.user.repos
}

export const getUserSkills = () => (state: RootState) => {
	return state.user.skills
}

export const getUserLoadingStatus = () => (state: RootState) => {
	return state.user.isLoading
}

export default userSlice.reducer
