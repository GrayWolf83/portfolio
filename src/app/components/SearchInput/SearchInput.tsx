import { Box, Button, TextField } from '@mui/material'
import React from 'react'
import { useAppDispatch } from '../../hooks/useAppReduxHooks'
import { searchUsersList } from '../../store/search'

interface ISearchInput {}

const SearchInput: React.FC<ISearchInput> = () => {
	const [searchValue, setSearchValue] = React.useState<string>('')
	const dispatch = useAppDispatch()

	const onChange = (
		e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
	) => {
		setSearchValue(e.target.value)
	}

	const handleSearch = (pageNumber: number) => {
		if (searchValue) {
			return dispatch(searchUsersList(searchValue, pageNumber))
		}
	}

	React.useEffect(() => {
		document.title = 'Поиск | GitHub view'
	}, [])

	return (
		<Box
			data-testid='SearchInput'
			sx={{
				width: '100%',
				display: 'flex',
				justifyContent: 'center',
				alignItems: 'flex-end',
			}}>
			<TextField
				label='Поиск'
				variant='standard'
				sx={{ width: '80%' }}
				placeholder='Введите имя пользователя'
				value={searchValue}
				onChange={onChange}
				required
			/>
			<Button
				variant='contained'
				size='medium'
				sx={{ ml: '5px' }}
				onClick={() => handleSearch(1)}>
				Найти
			</Button>
		</Box>
	)
}

export default SearchInput
