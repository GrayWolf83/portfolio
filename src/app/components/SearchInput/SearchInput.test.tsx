import React from 'react'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from '../../store'
import SearchInput from './SearchInput'

describe('SearchInput component', () => {
	test('SearchInput render', () => {
		render(
			<Provider store={store}>
				<BrowserRouter>
					<SearchInput />
				</BrowserRouter>
			</Provider>,
		)

		const searchInput = screen.getByTestId('SearchInput')
		expect(searchInput).toBeInTheDocument()
	})

	test('SearchInput snapshot', () => {
		expect(
			render(
				<Provider store={store}>
					<BrowserRouter>
						<SearchInput />
					</BrowserRouter>
				</Provider>,
			),
		).toMatchSnapshot()
	})
})
