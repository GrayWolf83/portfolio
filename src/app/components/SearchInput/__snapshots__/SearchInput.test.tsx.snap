// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component SearchInput snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1lp95k2"
        data-testid="SearchInput"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-w3s1g-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Поиск
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id=":r1:"
              placeholder="Введите имя пользователя"
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1yinwbo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Найти
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1lp95k2"
      data-testid="SearchInput"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-w3s1g-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r1:"
          id=":r1:-label"
        >
          Поиск
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-1ptx2yq-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r1:"
            placeholder="Введите имя пользователя"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1yinwbo-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Найти
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
